### PR TITLE
Build out Process discrepancy, Progressive, Turn out and Votes summary Reports

### LINKED ISSUE #304: Process discrepancy report: Create a List of polling stations and centers with forms in audit (based on region/ constituency/ sub constituency).
Build a table of polling stations and centres with result forms in audit state (based on region/ constituency/ sub constituency)

### PR DESCRIPTION
**Pull request by JohnMwashuma**
### Changes / Features implemented

- [x] Process discrepancy report by Regions, Constituency, Sub Constituency.
- [x] Progressive reports by Regions, Constituency, Sub Constituency.
- [x] Turnout reports by Regions, Constituency, Sub Constituency.
- [x] Votes summary reports by Regions, Constituency, Sub Constituency.

### Steps taken to verify this change does what is intended

- [x] Add/Update tests

### Side effects of implementing this change

Closes #304, #279, #284 , #308, #303, #273, #240, #227

### CODE REVIEW

File: tally_ho/apps/tally/views/data/center_list_view.py [hunk 1]
```diff
@@ -60,7 +75,24 @@ class CenterListView(LoginRequiredMixin,
 
     def get(self, *args, **kwargs):
         tally_id = kwargs.get('tally_id')
+        region_id = kwargs.get('region_id', None)
```
Comment (pld) [inline tally_ho/apps/tally/views/data/center_list_view.py:78]: the default is already `None`, I think it's fine to not pass it here and the below 2

File: tally_ho/apps/tally/views/data/center_list_view.py [hunk 2]
```diff
@@ -60,7 +75,24 @@ class CenterListView(LoginRequiredMixin,
...
+        constituency_id = kwargs.get('constituency_id', None)
+        sub_constituency_id = kwargs.get('sub_constituency_id', None)
         format_ = kwargs.get('format')
+        region_name = None
```
Comment (pld) [inline tally_ho/apps/tally/views/data/center_list_view.py:82]: optional, and I'm happy w/you not making this change

but I'd probably write this as

```python
region_name = region_id and Region.objects.get(id=region_id).name
...
```

and remove the conditionals below, it should be equivalent but less verbose

File: tally_ho/apps/tally/views/reports/administrative_areas_reports.py [hunk 3]
```diff
@@ -100,158 +323,483 @@ class RegionsReportsView(LoginRequiredMixin,
...
                               mixins.GroupRequiredMixin,
                               TemplateView):
     group_required = groups.TALLY_MANAGER
-    template_name = 'reports/administrative_areas_reports.html'
 
     def get(self, request, *args, **kwargs):
         tally_id = kwargs['tally_id']
-        export_type_ = kwargs.get('export_type')
+        region_id = kwargs['region_id']
+        report_type = kwargs.get('report_type', None)
```
Comment (pld) [inline tally_ho/apps/tally/views/reports/administrative_areas_reports.py:458]: same thing about default

File: tally_ho/apps/tally/views/reports/administrative_areas_reports.py [hunk 4]
```diff
@@ -100,158 +323,483 @@ class RegionsReportsView(LoginRequiredMixin,
...
+        constituency_id = kwargs.get('constituency_id', None)
+
+        region_name =\
```
Comment (pld) [inline tally_ho/apps/tally/views/reports/administrative_areas_reports.py:461]: think this is clearer with a conjunctive, and to be consistent I'm now thinking we should change the above

```python
region_name = region_id and Region.objects.get(...
```

File: tally_ho/apps/tally/views/reports/candidate_list_by_votes.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        sub_constituency_id = kwargs.get('sub_constituency_id', None)
+        ballot_report = self.request.session.get(
+            'ballot_report', None)
+
+        region_name = None
+        constituency_name = None
+        sub_constituency_code = None
+
+        if region_id:
+            region_name = Region.objects.get(id=region_id).name
```
Comment (pld) [inline tally_ho/apps/tally/views/reports/candidate_list_by_votes.py:90]: won't this (and below) error if the ID doesn't exist? that should be handled

File: tally_ho/apps/tally/views/data/center_list_view.py (hunk 1 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/data/center_list_view.py:78]: Makes sense

File: tally_ho/apps/tally/views/data/center_list_view.py (hunk 2 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/data/center_list_view.py:82]: Will do

File: tally_ho/apps/tally/views/reports/candidate_list_by_votes.py (hunk 5 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/reports/candidate_list_by_votes.py:90]: Will refactor this

File: tally_ho/apps/tally/static/js/data_table.js [hunk 6]
```diff
@@ -1,19 +1,40 @@
...
+      {
+        orderable: true,
+        searchable: true,
+        className: "center",
+        targets: [0, 1],
+      },
+    ],
+    searching: true,
+    processing: true,
+    serverSide,
```
Comment (pld) [inline tally_ho/apps/tally/static/js/data_table.js:19]: doesn't this key needs a value? it's failing the Codacy checks

File: tally_ho/apps/tally/static/js/data_table.js [hunk 7]
```diff
@@ -1,19 +1,40 @@
...
+    stateSave: true,
+    ajax: LIST_JSON_URL,
```
Comment (pld) [inline tally_ho/apps/tally/static/js/data_table.js:21]: not defined

File: tally_ho/apps/tally/static/js/data_table.js [hunk 8]
```diff
@@ -1,19 +1,40 @@
...
+    dom:
+      "<'row'<'col-sm-1'B><'col-sm-6'l><'col-sm-5'f>>" +
+      "<'row'<'col-sm-12'tr>>" +
+      "<'row'<'col-sm-5'i><'col-sm-7'p>>",
+    buttons: [
+      {
+        extend: "csv",
+        filename: exportFileName,
```
Comment (pld) [inline tally_ho/apps/tally/static/js/data_table.js:29]: not defined

-- if these are special flags for the library we can use override the style checks in-line

**Review by pld — CHANGES_REQUESTED** [review]

**Comment by JohnMwashuma** [comment]
@pld will respond to/address your comments once i have pushed some changes on these reports.

File: tally_ho/apps/tally/static/js/data_table.js (hunk 8 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/static/js/data_table.js:57]: These are special flags, I have ignore this file in the codacy config [here](https://github.com/onaio/tally-ho/blob/b854ebd638069b4843368cf217e1e500da24bfef/.codacy.yml#L4) and I can see it has been ignored in codacy [here](https://app.codacy.com/gh/onaio/tally-ho/settings/ignoreFiles) but it's still being captured.

File: tally_ho/apps/tally/views/reports/candidate_list_by_votes.py (hunk 5 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/reports/candidate_list_by_votes.py]: Done

File: tally_ho/apps/tally/views/reports/administrative_areas_reports.py (hunk 4 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/reports/administrative_areas_reports.py:2714]: Done

File: tally_ho/apps/tally/views/reports/administrative_areas_reports.py (hunk 3 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/reports/administrative_areas_reports.py:2711]: Done

File: tally_ho/apps/tally/views/data/center_list_view.py (hunk 2 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/data/center_list_view.py]: Done

File: tally_ho/apps/tally/views/data/center_list_view.py (hunk 1 above)
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/data/center_list_view.py]: Done

**Review by KipSigei — APPROVED** [review]

**Comment by JohnMwashuma** [comment]
@pld I have replaced travis with github actions since we are no longer running travis for public repos, please remove the tally project on travis and ignore the current travis status check.

**Review by pld — APPROVED** [review]

**Pull request merged**